File: Book_MS/src/main/java/ma/ensa/book_ms/Service/BookService.java
package ma.ensa.book_ms.Service;

import ma.ensa.book_ms.DTO.BookRequestDTO;
import ma.ensa.book_ms.DTO.BookResponseDTO;
import java.lang.Long;

import java.util.List;

public interface BookService {

    BookResponseDTO saveBook(BookRequestDTO bookRequestDTO);

    BookResponseDTO getBookById(Long id);

    List<BookResponseDTO> getAllBooks();

    List<BookResponseDTO> getAvailableBooks();

    BookResponseDTO incrementAvailableCopies(Long id);

    BookResponseDTO decrementAvailableCopies(Long id);

    void deleteBook(Long id);

    List<BookResponseDTO> searchByTitle(String keyword);

    List<BookResponseDTO> searchByAuthor(String keyword);

    List<BookResponseDTO> getBooksByIds(List<Long> ids);

}

File: frontend/src/app/core/services/book.service.ts
// src/app/core/services/book.service.ts

import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { BookRequest, BookResponse } from '../models/models';
import {environment} from '../../../environments/environment';
@Injectable({
  providedIn: 'root'
})
export class BookService {
  private readonly API_URL =`${environment.apiUrl}/book-ms/api/books`;
  constructor(private http: HttpClient) {}

  // Create book (ADMIN only)
  createBook(book: BookRequest): Observable<BookResponse> {
    return this.http.post<BookResponse>(this.API_URL, book);
  }

  // Get all books
  getAllBooks(): Observable<BookResponse[]> {
    return this.http.get<BookResponse[]>(this.API_URL);
  }

  // Get available books
  getAvailableBooks(): Observable<BookResponse[]> {
    return this.http.get<BookResponse[]>(`${this.API_URL}/available`);
  }

  // Get book by ID
  getBookById(id: number): Observable<BookResponse> {
    return this.http.get<BookResponse>(`${this.API_URL}/${id}`);
  }

  // Delete book (ADMIN only)
  deleteBook(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  // Search by title
  searchByTitle(keyword: string): Observable<BookResponse[]> {
    const params = new HttpParams().set('keyword', keyword);
    return this.http.get<BookResponse[]>(`${this.API_URL}/search/title`, { params });
  }

  // Search by author
  searchByAuthor(keyword: string): Observable<BookResponse[]> {
    const params = new HttpParams().set('keyword', keyword);
    return this.http.get<BookResponse[]>(`${this.API_URL}/search/author`, { params });
  }
}
